experimental/Schemes/singular_locus.jl

```diff
@@ -0,0 +1,147 @@
+export is_equidimensional
+export singular_locus
+export reduced_scheme
+export is_smooth
+
+function singular_locus(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  comp = _singular_locus_with_decomposition(X)
+  if length(comp) == 0 
+    return subscheme(X, one(OO(X)))
+  end
+  R = base_ring(OO(X))
+  return Spec(R, prod([modulus(OO(Y)) for Y in comp]), inverted_set(OO(X)))
+end
+
+function _singular_locus_with_decomposition(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  I = localized_modulus(OO(X))
+  result = typeof(X)[]
+  if is_equidimensional(X)
+    d = dim(X)
+    R = base_ring(OO(X))
+    n = ngens(R)
+    J = saturated_ideal(I)
+    M = jacobi_matrix(gens(J))
+    K = ideal(R, minors(M, n-d))
+    one(OO(X)) in OO(X)(K) && return result
+    return [subscheme(X, K)]
+  else
+    P = primary_decomposition(saturated_ideal(I))
+    components = [subscheme(X, J[2]) for J in P]
+    for Y in components
+      for Z in components
+        if !(Y == Z)
+          W = intersect(Y, Z)
+          if !isempty(W) 
+            push!(result, W)
+          end
+        end
+      end
+    end
+    for Y in components
+      result = vcat(result, singular_locus(Y))
+    end
+  end
+  return result
+end
+
+@attr Bool function is_equidimensional(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  I = localized_modulus(OO(X))
+  P = primary_decomposition(saturated_ideal(I))
+  length(P) == 0 && return true
+  d = dim(P[1][1])
+  all(x->dim(x[1])==d, P[2:end]) && return true
+  return false
+end
+
+@attr function reduced_scheme(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  I = localized_modulus(OO(X))
+  J = radical(saturated_ideal(I))
+  return Spec(base_ring(J), J, inverted_set(OO(X)))
+end
+
+@attr function is_reduced(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  I = localized_modulus(OO(X))
+  J = saturated_ideal(I)
+  return is_reduced(quo(base_ring(OO(X)), J)[1])
+end
+
+
+########################################################################
+# Additional functionality for fractions                               #
+########################################################################
+
+function derivative(f::MPolyLocalizedRingElem, i::Int)
+  num = derivative(numerator(f), i)*denominator(f) - derivative(denominator(f), i)*numerator(f)
+  den = denominator(f)^2
+  g = gcd(num, den)
+  return parent(f)(divexact(num, g), divexact(den, g), check=false)
+end
+
+function derivative(f::MPolyQuoLocalizedRingElem, i::Int)
+  num = derivative(lifted_numerator(f), i)*lifted_denominator(f) - derivative(lifted_denominator(f), i)*lifted_numerator(f)
+  den = lifted_denominator(f)^2
+  g = gcd(num, den)
+  return parent(f)(divexact(num, g), divexact(den, g), check=false)
+end
+
+function jacobi_matrix(f::MPolyLocalizedRingElem)
```

Comment:
General question on jacobian matrix (boiling down to derivative computation) in localized rings: I agree that the exact functionality needs to be available, but programmer-facing it might be a good idea to also provide this in a more lazy version omitting the denominators from the beginning whenever a suitable flag is set. I am thinking of contexts where subsequent computations will be performed modulo the ideal in question and one does not want to first introduce extra terms just to delete them later on. This might be a flag like forget_denominators=false among the arguments..

Reply:
No, I clearly object to that. If we wish to do things in localized rings, we need to be sure to have the corresponding elements. Otherwise, it becomes completely messy to write generic code which is parametrized with the type of the ring. If anything, we would have to make elements of localized rings behave like polynomials whenever they effectively are polynomials; e.g. print without denominator, etc. Starting implementations with flags is not user-friendly and breaks type stability. Since addition, multiplication, etc. is critical, type stability is key here.